src/ucp/core/ucp_listener.c

```diff
@@ -172,99 +172,239 @@ static void ucp_listener_conn_request_callback(uct_iface_h tl_iface, void *arg,
-                           UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER |
-                           UCP_LISTENER_PARAM_FIELD_CONN_HANDLER)) {
-        ucs_error("Only one accept handler should be provided");
-        return UCS_ERR_INVALID_PARAM;
+    listener->cms.num_listeners = 0;
+    uct_listeners               = ucs_calloc(num_cms, sizeof(*uct_listeners),
+                                             "uct_listeners_arr");
+    if (uct_listeners == NULL) {
+        ucs_error("Can't allocate memory for UCT listeners array");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    UCS_ASYNC_BLOCK(&worker->async);
+    listener->cms.listeners = uct_listeners;
 
-    listener = ucs_calloc(1, sizeof(*listener), "ucp_listener");
-    if (listener == NULL) {
-        status = UCS_ERR_NO_MEMORY;
-        goto out;
-    }
+    for (i = 0; i < num_cms; ++i) {
+        status = uct_listener_create(worker->cms[i], &addr,
+                                     params->sockaddr.addrlen, &uct_params,
+                                     &uct_listeners[listener->cms.num_listeners]);
+        if (status != UCS_OK) {
+            ucs_debug("failed to create UCT listener on CM %p with address %s status %s",
+                      worker->cms[i],
+                      ucs_sockaddr_str(params->sockaddr.addr, addr_str,
+                                       UCS_SOCKADDR_STRING_LEN),
+                      ucs_status_string(status));
+            continue;
+        }
 
-    listener->worker = worker;
+        ++listener->cms.num_listeners;
 
-    if (params->field_mask & UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER) {
-        UCP_CHECK_PARAM_NON_NULL(params->accept_handler.cb, status,
-                                 goto err_free_listener);
-        listener->accept_cb = params->accept_handler.cb;
-        listener->arg       = params->accept_handler.arg;
-    } else if (params->field_mask & UCP_LISTENER_PARAM_FIELD_CONN_HANDLER) {
-        UCP_CHECK_PARAM_NON_NULL(params->conn_handler.cb, status,
-                                 goto err_free_listener);
-        listener->conn_cb   = params->conn_handler.cb;
-        listener->arg       = params->conn_handler.arg;
+        status = ucs_sockaddr_get_port(&addr, &port);
+        if (status != UCS_OK) {
+            goto destroy_listeners;
+        }
+
+        uct_attr.field_mask = UCT_LISTENER_ATTR_FIELD_SOCKADDR;
+        status = uct_listener_query(uct_listeners[listener->cms.num_listeners - 1],
+                                    &uct_attr);
+        if (status != UCS_OK) {
+            goto destroy_listeners;
+        }
+
+        status = ucs_sockaddr_get_port((struct sockaddr *)&uct_attr.sockaddr,
+                                       &listener_port);
+        if (status != UCS_OK) {
+            goto destroy_listeners;
+        }
+
+        if (port != listener_port) {
+            ucs_assert(port == 0);
+            status = ucs_sockaddr_set_port(&addr, listener_port);
```

Comment:
maybe just save the port number in the UCP listener.. will avoid all this complex logic of listener_query at the cost of uint16_t

Reply:
this complex logic solves a different problem - when user don't specify the port in sockaddr.

Reply:
but ucp_listen_on_cm() queries the port anyway? so you should have the port number even if the user doesn't specify it

Reply:
yes, but only after 1st iteration, then 1st UCT listener is started.